content/blog/2022-05-18-learned-federated-query-optimization.md

```diff
@@ -0,0 +1,178 @@
+heterogeneity of the underlying infrastructure and it's able to optimize the query with respect to each table location.
+
+# What did you do?
+We mainly focused on the optimizer of our federated query compiler. We implemented a federated query compiler which is
+able to do optimize federated queries by splitting the execution of the initial query into individual, smaller parts
+that are pushed-down to the external database systems, achieving significant performance improvements. The optimizer
+of our query compiler is based on a machine learning approach which treats the external systems as black-boxes,
+profiles them and learns how to estimate the query execution cost of each individual system. Thus, our federated query
+compiler is able to easily connect with any system that supports SQL, without needing any expertise of the underlying 
+infrastructure.
+
+## The Self-Learned Federated Query Compiler
+We developed a query compiler that optimizes input queries with respect to the underlying table locations. To explain 
+how our optimizer works, let's take a look at the following example. Consider the query:
+
+```sql
+SELECT AVG(A.salary)
+FROM A, B, C
+WHERE A.id = B.id
+AND A.id = C.id
+AND B.id = C.id
+```
+This query consists of a join between three tables: A, B and C. Let's assume three different data sources, namely DS1 
+and DS2 and assume the following table locations:
+
+In relational algebra, this join would look like the following: 
+`A ⋈ B ⋈ C`. By default, a federation engine (like Spark SQL) would first transfer all tables A, B and C to its system
+and then, it would execute the query its execution engine. However, more complex strategies could be followed for a query 
+like this. For example, `A ⋈ B ⋈ C` could be split into two complement subqueries, `A` (scan) and `B ⋈ C` (join), where 
+the first query which is just a scan would be executed as a simple scan from DS1. However, the second query (`B ⋈ C`) 
+could be pushed-down for local execution in `DS2`. But how do we know if pushing-down `B ⋈ C` to `DS2` would result to
+a better performance compared to just fetching tables `B` and `C` to the federation engine (Spark SQL) and executing the 
+join there?
+
+## Learned Query Compilation
+To answer the previous question, we use a machine learning model as our cost model that helps us evaluate the cost of 
+a query without executing it. Using this model, we can easily evaluate whether it is better to push-down a query or not.
+In order to do that, we use a unified vector space model that is able to transform SQL queries into vectors, and thus,
+we can feed these queries to our model in order to train and predict the execution time of future queries. Thus, our
+optimizer is able to leverage this machine learning model in order to decide which parts of the query should push-down 
+or not.
```

Comment:
Two important questions this paragraph might raise:

* What is a "unified vector space model"? Is it a pre-trained embedding, like word2vec but for SQL queries?
* This is supervised learning, right? Like, the learned function is from SQL queries to latencies (in some specific unit, like seconds)? If so, when/how will it be trained (which is relevant to knowing when it will need to be retrained—is it every change to the data?)?